Add support for "files"-archive

bbilger opened the pull request:
i.e. write individual pbf-files to disk in the format BASE/z/x/y.pbf

in order to use that format it must be passed as "--ouput=/path/to/tiles?format=files"

Fixes #536

note: it also supports concurrent writing - just have to pass --tile_write_threads=5

msbarry commented on planetiler-core/src/main/java/com/onthegomap/planetiler/files/WriteableFilesArchive.java:
```diff
@@ -0,0 +1,92 @@
+    // nothing to do here
+  }
+
+  private static void mkdirs(Path p) {
+    try {
+      Files.createDirectories(p);
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
+    }
+  }
```
You could use `FileUtils.createDirectory` as well, which does the same catch/throw.

msbarry commented on planetiler-core/src/test/java/com/onthegomap/planetiler/archive/TileArchiveConfigTest.java:
```diff
@@ -33,4 +42,51 @@ void testPmtiles() {
+  @Test
+  void testExistsForFilesArchive(@TempDir Path tempDir) throws IOException {
+    final Path out = tempDir.resolve("outdir");
+    final var config = TileArchiveConfig.from(out + "?format=files");
```
Currently when `format=` is missing `TileArchiveConfig` defaults to mbtiles. I wonder if when the output path has no extension or ends with a slash if it should default to `files` as a small ease of use improvement?

bbilger replied:
done https://github.com/onthegomap/planetiler/pull/761/commits/1faf56611a11f9e92ec02f631262689c8a13e417

msbarry commented on planetiler-core/src/main/java/com/onthegomap/planetiler/files/WriteableFilesArchive.java:
```diff
@@ -0,0 +1,92 @@
+      }
+      lastCheckedFolder = folder;
+      try {
+        Files.write(file, encodingResult.tileData());
```
`encodingResult.tileData()` is gzipped here. We should probably make it configurable whether or not the output files should be gzipped (I think default to false?). If they aren't gzipped we can skip the gzip step while encoding, and if they are gzipped the output extension should probably be `.pbf.gz`

bbilger replied:
The output is not necessarily gzipped. This is already configurable at a higher level: tile_compression. So I don't think anything needs to be handled here specifically at the archive level.

bbilger replied:
I personally prefer the tiles to be at a known location by default. With the introduction of the tile_scheme param users can adjust as needed.
https://github.com/onthegomap/planetiler/commit/2e4652ef39e501b472265852c261d7685cf179ee

msbarry commented:
Regarding performance, this is definitely going to be slower but I wonder if there's any performance difference to `Files.write` from multiple OS threads vs. `Files.write` from virtual threads vs. something like `AsynchronousFileChannel` (or others?).

These could also be separate followup issues, but I noticed:
- when there are multiple writer threads with `--tile_write_threads=5` they each print `Finished z11 in 0.4s cpu:0.6s avg:1.7, now starting z12`, call `tileWriter.printStats()` and call `archive.finish()` - maybe there should be a different way to handle those? Or only do the logging from thread 0 or something?
- `FileUtils.size` looks like a bottleneck because the `ProgressLoggers` calls it while printing status during the archive write phase. Might make sense to keep a counter of bytes written to the output directory instead of having to scan it each time?

bbilger commented on planetiler-core/src/main/java/com/onthegomap/planetiler/files/FilesArchiveUtils.java:
```diff
@@ -0,0 +1,21 @@
+  static Path relativePathFromTileCoord(TileCoord tc) {
+    return Paths.get(Integer.toString(tc.z()), Integer.toString(tc.x()),
+      tc.y() + PBF_FILE_ENDING);
+  }
```
added support for custom "tile_scheme" supported templates are `{x}`, `{y}`, `{z}` as well as `{xs}` and `{ys}` (s=safe). `{xs}` and `{ys}` are split into 2 levels.
```
{z}/{x}/{y}.pbf     => 3/1/2.pbf
{x}/{y}/{z}.pbf     => 1/2/3.pbf
{x}-{y}-{z}.pbf     => 1-2-3.pbf
{x}/a/{y}/b{z}.pbf  => 1/a/2/b3.pbf
{z}/{x}/{y}.pbf.gz  => 3/1/2.pbf.gz
{z}/{xs}/{ys}.pbf   => 3/000/001/000/002.pbf
{z}/{x}/{ys}.pbf    => 3/1/000/002.pbf
{z}/{xs}/{y}.pbf    => 3/000/001/2.pbf
```

bbilger commented on planetiler-core/src/main/java/com/onthegomap/planetiler/files/ReadableFilesArchive.java:
```diff
@@ -0,0 +1,99 @@
+  @Override
+  public TileArchiveMetadata metadata() {
+    return null;
+  }
```
done

did a bunch of refactorings to the (de-)serialization of TileArchiveMetadata

https://github.com/onthegomap/planetiler/pull/761/commits/d238a4f30ca63abc450a4c7361a11705a13f501c

bbilger commented:
> These could also be separate followup issues, but I noticed:
> 
> * when there are multiple writer threads with `--tile_write_threads=5` they each print `Finished z11 ... now starting z12`, call `tileWriter.printStats()` and call `archive.finish()` - maybe there should be a different way to handle those? Or only do the logging from thread 0 or something?
> * `FileUtils.size` looks like a bottleneck because the `ProgressLoggers` calls it while printing status during the archive write phase, and `Stats` calls it at the end to print a file size summary. Might make sense to either disable (have the method give up a certain number of files or for a certain amount of time) or keep a counter of bytes written to the output directory instead of having to scan it each time?

Addressed those points here: https://github.com/onthegomap/planetiler/pull/761/commits/5371520256376a3b61114d54b0e56be75e19bf5d

1. `archive.finish()` being called for each writer was definitely a bug - it's now called after ALL writers finished
2. `Finished z11 ... now starting z12`, call `tileWriter.printStats()` - logging for the first writer only - not 100% accurate but should be good enough for some "zoom-progress logging"
3. `tileWriter.printStats()` think that's ok that we print stats for each writer as long as they are writer specific. I introduced printStats on WriteableTilesArchive now for archive related logging.
4. `FileUtils.size` in `ProgressLoggers`  was definitely an issue. I introduced WriteableTilesArchive#bytesWritten in order to handle that correctly (was also an issue for the stream writers).

bbilger commented:
> Regarding performance, this is definitely going to be slower than a single-file archive but I wonder if there's any performance difference to `Files.write` from multiple OS threads vs. `Files.write` from virtual threads vs. something like `AsynchronousFileChannel` (or others?).

Good question/point. Need to investigate a bit on this.

bbilger commented:
regarding performance: I wrote some benchmark (BenchmarkTileFilesWrites) that loads a mbtiles file into memory and then writes the tiles to files in different modes.

See the benchmark results below from my machine for US (0<=z<=13). If I interpret things correctly, then there's no real benefit for virtual threads. At the same time async seems to be good compromise between absolute and CPU time.

So I impletemented:
1. async file writing for the files archive with the option to write sync.
2. a way to use virtual threads for the tile writers (`--tile_write_threads=100 --tile_write_executor=virtual`) - in case you want to experiment with it

If you agree, then I'd remove the benchmark as well as the option to use virtual threads for writing, again. 

| mode | run | duration |
| --- | --- | --- |
| fixed1 | 0 | 4m34s cpu:4m25s gc:1s avg:1 |
| fixed1 | 1 | 4m20s cpu:4m21s gc:2s avg:1 |
| fixed2 | 0 | 3m12s cpu:5m56s avg:1.9 |
| fixed2 | 1 | 2m54s cpu:5m43s avg:2 |
| fixed4 | 0 | 2m55s cpu:11m2s avg:3.8 |
| fixed4 | 1 | 2m56s cpu:11m2s avg:3.8 |
| fixed10Virtual | 0 | 3m18s cpu:22m21s avg:6.8 |
| fixed10Virtual | 1 | 3m19s cpu:22m30s avg:6.8 |
| fixed100Virtual | 0 | 3m30s cpu:30m14s gc:1s avg:8.6 |
| fixed100Virtual | 1 | 3m33s cpu:30m32s avg:8.6 |
| fixed1000Virtual | 0 | 3m42s cpu:33m47s gc:1s avg:9.1 |
| fixed1000Virtual | 1 | 3m46s cpu:33m56s avg:9 |
| async | 0 | 3m53s cpu:6m56s avg:1.8 |
| async | 1 | 3m54s cpu:7m13s avg:1.8 |
| unboundVirtual | 0 | 4m29s cpu:19m42s gc:3s avg:4.4 |
| unboundVirtual | 1 | 4m37s cpu:20m34s gc:3s avg:4.5 |

changes are here: https://github.com/onthegomap/planetiler/pull/761/commits/b8cfa56977d98520aa8b62252c3a2726d440afe0

msbarry commented:
> regarding performance: I wrote some benchmark (BenchmarkTileFilesWrites) that loads a mbtiles file into memory and then writes the tiles to files in different modes.
> 
> See the benchmark results below from my machine for US (0<=z<=13). If I interpret things correctly, then there's no real benefit for virtual threads. At the same time async seems to be good compromise between absolute and CPU time.
> 
> So I impletemented:
> 
> 1. async file writing for the files archive with the option to write sync.
> 2. a way to use virtual threads (bound only) for the tile writers (`--tile_write_threads=100 --tile_write_executor=virtual`) - in case you want to experiment with it
> 
> If you agree, then I'd remove the benchmark as well as the option to use virtual threads for writing, again.
> 
> mode	run	duration
> fixed1	0	4m34s cpu:4m25s gc:1s avg:1
> fixed1	1	4m20s cpu:4m21s gc:2s avg:1
> fixed2	0	3m12s cpu:5m56s avg:1.9
> fixed2	1	2m54s cpu:5m43s avg:2
> fixed4	0	2m55s cpu:11m2s avg:3.8
> fixed4	1	2m56s cpu:11m2s avg:3.8
> fixed10Virtual	0	3m18s cpu:22m21s avg:6.8
> fixed10Virtual	1	3m19s cpu:22m30s avg:6.8
> fixed100Virtual	0	3m30s cpu:30m14s gc:1s avg:8.6
> fixed100Virtual	1	3m33s cpu:30m32s avg:8.6
> fixed1000Virtual	0	3m42s cpu:33m47s gc:1s avg:9.1
> fixed1000Virtual	1	3m46s cpu:33m56s avg:9
> async	0	3m53s cpu:6m56s avg:1.8
> async	1	3m54s cpu:7m13s avg:1.8
> unboundVirtual	0	4m29s cpu:19m42s gc:3s avg:4.4
> unboundVirtual	1	4m37s cpu:20m34s gc:3s avg:4.5
> changes are here: [b8cfa56](https://github.com/onthegomap/planetiler/commit/b8cfa56977d98520aa8b62252c3a2726d440afe0)

Thanks for doing those tests! I saw similar results running this on a linux box with fast SSD - fastest seemed to be with 4 dedicated OS threads.

Given these results, can we just remove the extra code paths and keep the configurable number of fixed threads you had initially since it's the fastest and simplest?

msbarry commented on planetiler-core/src/main/java/com/onthegomap/planetiler/files/WriteableFilesArchive.java:
```diff
@@ -0,0 +1,250 @@
+    }
+  }
+
+  private static class AsyncTileFilesWriter extends BaseTileFilesWriter {
```
Let's remove the async tile writer and collapse into a single sync writer class here for simplicity

msbarry commented on planetiler-core/src/main/java/com/onthegomap/planetiler/stats/ProcessInfo.java:
```diff
@@ -234,6 +234,7 @@ public static Map<Long, ThreadState> getThreadStats() {
     ThreadMXBean threadMXBean = ManagementFactory.getThreadMXBean();
     ThreadInfo thread = threadMXBean.getThreadInfo(Thread.currentThread().getId());
-    return new ThreadState(threadMXBean, thread);
+    // null for virtual - tmp workaround
+    return thread == null ? null : new ThreadState(threadMXBean, thread);
```
We can remove this too.

msbarry commented on planetiler-core/src/main/java/com/onthegomap/planetiler/stats/Timers.java:
```diff
@@ -115,7 +115,11 @@ public Finishable startTimer(String name, boolean logStart) {
     Stage stage = currentStage.get();
     if (stage != null) {
-      stage.threadStats.add(new ThreadInfo(ProcessInfo.getCurrentThreadState(), prefix, elapsed));
+      final ProcessInfo.ThreadState threadState = ProcessInfo.getCurrentThreadState();
+      // null for virtual - tmp workaround
+      if (threadState != null) {
+        stage.threadStats.add(new ThreadInfo(threadState, prefix, elapsed));
+      }
```
We can remove this

msbarry commented on planetiler-core/src/main/java/com/onthegomap/planetiler/stream/WriteableStreamArchive.java:
```diff
@@ -34,6 +36,8 @@
  */
 abstract class WriteableStreamArchive implements WriteableTileArchive {
 
+  private final LongAdder bytesWritten = new LongAdder();
```
Can we use `Counter.newMultiThreadCounter` here and get a new `counterForThread` from it in the factory?

msbarry commented on planetiler-core/src/main/java/com/onthegomap/planetiler/worker/Worker.java:
```diff
@@ -41,6 +41,20 @@ public Worker(String prefix, Stats stats, int threads, RunnableThatThrows task)
+   * @param task         the work to do in each thread
+   */
+  @SuppressWarnings("java:S1181")
+  public Worker(String prefix, Stats stats, int threads, ExecutorType executorType, RunnableThatThrows task) {
+    this(prefix, stats, threads, executorType, i -> task.run());
+  }
```
Let's remove this as well since we're just doing the fixed threads. I think things using virtual threads going forward will probably use a slightly different paradigm (new virtual thread per task, with a semaphore to limit concurrency) so probably don't need to be supported in the worker framework.

msbarry commented on planetiler-core/src/main/java/com/onthegomap/planetiler/archive/WriteableTileArchive.java:
```diff
@@ -44,6 +44,10 @@ default void initialize() {}
 
+  default void printStats() {}
+
+  long bytesWritten();
```
Thanks for adding this, I could do this in a separate PR as well but there's one remaining place where it gets the size of the archive which could take a very long time if you just wrote tiles for the planet: `Stats.printSummary` - ideally it either uses the same output from this method, or we add a failsafe to `FileUtils.size` that gives up after a certain number of files/time since that output duplicates what got printed earlier we won't miss much.

bbilger replied:
done: https://github.com/onthegomap/planetiler/pull/761/commits/c41b95354f224a112cba09e380ccac9de8aa235b

msbarry reviewed (commented):
Thanks for making these changes! I'm trying to test it with the custom schemes but getting some errors.

My first attempt was to use `--output=tiles/{z}/{x}-{y}.pbf.gz` - it would be cool if that worked but looks like it's not currently handled.

Then I tried `--output=tiles?format=files&tile_scheme={z}/{x}/{y}.pbf.gz` but I get `Illegal character in query at index 31: tiles?format=files&tile_scheme={z}/{x}/{y}.pbf.gz`.

To make it work I had to do `--output=tiles&format=files&tile_scheme=%7Bz%7D%2F%7Bx%7D%2F%7By%7D.pbf.gz` which isn't super easy to use.

Is there another way that I'm missing?

bbilger commented:
> Thanks for making these changes! I'm trying to test it with the custom schemes but getting some errors:
> 
> My first attempt was to use `--output=tiles/{z}/{x}-{y}.pbf.gz` - it would be cool if that worked but looks like it's not currently handled.
> 
> Then I tried `--output=tiles?format=files&tile_scheme={z}/{x}/{y}.pbf.gz` but I get `Illegal character in query at index 31: tiles?format=files&tile_scheme={z}/{x}/{y}.pbf.gz`.
> 
> To make it work I had to do `--output=tiles&format=files&tile_scheme=%7Bz%7D%2F%7Bx%7D%2F%7By%7D.pbf.gz` which isn't super user-friendly.
> 
> Is there another way that I'm missing?

The other option - which I always use - is to use format pre-fixed args => `--output=tiles/ --files_tile_scheme={x}/{y}/{z}.pbf`

I'll check if we can support your other approaches.

msbarry commented:
> > Thanks for making these changes! I'm trying to test it with the custom schemes but getting some errors:
> > My first attempt was to use `--output=tiles/{z}/{x}-{y}.pbf.gz` - it would be cool if that worked but looks like it's not currently handled.
> > Then I tried `--output=tiles?format=files&tile_scheme={z}/{x}/{y}.pbf.gz` but I get `Illegal character in query at index 31: tiles?format=files&tile_scheme={z}/{x}/{y}.pbf.gz`.
> > To make it work I had to do `--output=tiles&format=files&tile_scheme=%7Bz%7D%2F%7Bx%7D%2F%7By%7D.pbf.gz` which isn't super user-friendly.
> > Is there another way that I'm missing?
> 
> The other option - which I always use - is to use format pre-fixed args => `--output=tiles/ --files_tile_scheme={x}/{y}/{z}.pbf`
> 
> I'll check if we can support your other approaches.

Ah got it, that makes sense.  Could we add a short javadoc to the major new classes and maybe include that cli arg snippet in the javadoc for readable/writeable files archive classes?

The shorthand one would be a nice to have, but might be hard to disambiguate between that and writing to a file archive on disk?

bbilger commented on planetiler-core/src/main/java/com/onthegomap/planetiler/archive/TileArchiveMetadataDeSer.java:
```diff
@@ -0,0 +1,239 @@
+/**
+ * Container for everything related to (de-)serialization of {@link TileArchiveMetadata}
+ */
+public final class TileArchiveMetadataDeSer {
```
np - I think overall the (de-) serialization is now more centralized and less duplicate code

bbilger commented on planetiler-core/src/main/java/com/onthegomap/planetiler/archive/TileArchiveWriter.java:
```diff
@@ -371,8 +387,6 @@ private void tileWriter(Iterable<TileBatch> tileBatches) throws ExecutionExcepti
     if (time != null) {
```
I think there's still a point in having printStats for a writer. And there's one usage https://github.com/onthegomap/planetiler/blob/cbb092a1f76925e0642cc4c8a954783f59c9f3aa/planetiler-core/src/main/java/com/onthegomap/planetiler/mbtiles/Mbtiles.java#L804
Moving that to the archive's printStats feels strange. But if we for sure could if you want, or log from #close in that specific case.

bbilger replied:
I am not quite sure what you mean here, sorry

bbilger commented on planetiler-core/src/main/java/com/onthegomap/planetiler/archive/WriteableTileArchive.java:
```diff
@@ -44,6 +44,10 @@ default void initialize() {}
    */
   default void finish(TileArchiveMetadata tileArchiveMetadata) {}
 
+  default void printStats() {}
```
I think there's still a point in having printStats for a writer. And there's one usage https://github.com/onthegomap/planetiler/blob/cbb092a1f76925e0642cc4c8a954783f59c9f3aa/planetiler-core/src/main/java/com/onthegomap/planetiler/mbtiles/Mbtiles.java#L804
Moving that to the archive's printStats feels strange. But if we for sure could if you want, or log from #close in that specific case.

bbilger commented on planetiler-core/src/main/java/com/onthegomap/planetiler/files/FilesArchiveUtils.java:
```diff
@@ -0,0 +1,56 @@
+    return new TileSchemeEncoding(tileScheme, basePath);
+  }
+
+  static WriteableFilesArchive.WriteMode writeMode(Arguments options) {
+    final String mode = options.getString(
+      OPTION_WRITE_MODE,
+      "the write mode - either sync or async",
+      WriteableFilesArchive.WriteMode.ASYNC.toString().toLowerCase()
+    );
+    return WriteableFilesArchive.WriteMode.valueOf(mode.toUpperCase());
+  }
+}
```
there's not a lot code for the async mode, and it might be beneficial to have it but okay, I removed it again

bbilger commented on planetiler-core/src/main/java/com/onthegomap/planetiler/archive/TileArchiveWriter.java:
```diff
@@ -330,11 +334,18 @@ private void tileEncoderSink(Iterable<TileBatch> prev) throws IOException {
+    final boolean firstTileWriter = firstTileWriterTracker.compareAndExchange(true, false);
+    if (firstTileWriter) {
+      archive.initialize();
+    }
```
Good point, was thinking about it as well but thought not very likely.
Didn't find any good spot to move it to. I moved it here now: https://github.com/onthegomap/planetiler/blob/e39551859bffc08d6b2f769f7cdf47bb20ccd614/planetiler-core/src/main/java/com/onthegomap/planetiler/archive/TileArchiveWriter.java#L154
Let me know if you have a better idea.

bbilger commented:
> > > Thanks for making these changes! I'm trying to test it with the custom schemes but getting some errors:
> > > My first attempt was to use `--output=tiles/{z}/{x}-{y}.pbf.gz` - it would be cool if that worked but looks like it's not currently handled.
> > > Then I tried `--output=tiles?format=files&tile_scheme={z}/{x}/{y}.pbf.gz` but I get `Illegal character in query at index 31: tiles?format=files&tile_scheme={z}/{x}/{y}.pbf.gz`.
> > > To make it work I had to do `--output=tiles&format=files&tile_scheme=%7Bz%7D%2F%7Bx%7D%2F%7By%7D.pbf.gz` which isn't super user-friendly.
> > > Is there another way that I'm missing?
> > 
> > 
> > The other option - which I always use - is to use format pre-fixed args => `--output=tiles/ --files_tile_scheme={x}/{y}/{z}.pbf`
> > I'll check if we can support your other approaches.
> 
> Ah got it, that makes sense. Could we add a short javadoc to the major new classes (don't worry about all of them, the helper classes should be self-explanatory without any javadoc) and maybe include that cli arg snippet in the javadoc for readable/writeable files archive classes?
> 
> The shorthand one would be a nice to have, but might be hard to disambiguate between that and writing to a file archive on disk?

I added support for those shortcuts here: https://github.com/onthegomap/planetiler/pull/761/commits/7d0bc26b72d1ee7b95c85d8a5196cbbecf7040d4
Detecting it is rather simple by the templates ({) - which otherwise should not really be used, and even if so users can pass the format-query param to enforce some format.

I also added javadoc where I thought it makes sense but let me know if you want more.

msbarry reviewed (commented):
Thanks for making those changes, it's looking good, a couple of minor comments...

bbilger commented on planetiler-core/src/main/java/com/onthegomap/planetiler/files/FilesArchiveUtils.java:
```diff
@@ -0,0 +1,73 @@
+
+  static BasePathWithTileSchemeEncoding basePathWithTileSchemeEncoding(Arguments options, Path basePath) {
+    final String basePathStr = basePath.toString();
+    final int curlyIndex = basePathStr.indexOf('{');
```
good catch! adjusted the implementation accordingly

bbilger commented:
Should be all set now. Let me know if there's anything else.

msbarry reviewed (approved):
Looks great, thank you!!
